### PR TITLE
Add `#![rustc_never_type_mode = "..."]` crate-level attribute to allow experimenting

### PR DESCRIPTION
**Pull request by WaffleLapkin**
Demonstrating how different approaches with the never type work is somewhat hard when you can't actually provide a runnable example. Let's add features that change the fallback behavior.

This adds `#![rustc_never_type_mode = "no_fallback"]` and `#![rustc_never_type_mode = "fallback_to_never"]`, but I also plan to add others (in future PRs). 

cc @traviscross 
r? @compiler-errors

### CODE REVIEW

**Comment by compiler-errors** [comment]
I am having trouble finding value in adding these as feature flags. I think they should just be crate-level `![rustc_never_type_fallback = ".."]` flags, since I don't believe we ever expect to allow users to choose this behavior.

**Comment by compiler-errors** [comment]
Especially if you plan to add *more* feature gates, this just seems really unmanageable.

**Comment by WaffleLapkin** [comment]
Well, yes, this is more for testing stuff and making sure T-lang understands what is the proposal with the never type.

@compiler-errors do you have a pointer of where a crate-level attribute was added? I made this as feature(s) because it was simpler to do, but I can rewrite this as an attribute, just not sure where those are handled in rustc..

**Comment by compiler-errors** [comment]
We don't really "handle" crate-level attributes anywhere centralized.

They are just checked with `tcx.has_attr(CRATE_DEF_ID, ..)`. Check out `rustc_hidden_type_of_opaques` for example. I guess use `get_attr` so you can read the meta part of the attribute.

**Comment by WaffleLapkin** [comment]
@compiler-errors thanks <3 made this into an attribute.

**Comment by WaffleLapkin** [comment]
There is a place where I'm not sure what to do:

https://github.com/rust-lang/rust/blob/c5b571310dc60c6a58c6505cce2fb7e2f3f9aa68/compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs#L396-L402

I have no context on what this does..

File: compiler/rustc_feature/src/builtin_attrs.rs [hunk 1]
```diff
@@ -580,6 +580,13 @@ pub const BUILTIN_ATTRIBUTES: &[BuiltinAttribute] = &[
         "`may_dangle` has unstable semantics and may be removed in the future",
     ),
 
+    rustc_attr!(
+        rustc_never_type_mode, Normal, template!(NameValueStr: "fallback_to_unit|fallback_to_niko|no_fallback"), ErrorFollowing,
```
Comment (compiler-errors) [inline compiler/rustc_feature/src/builtin_attrs.rs:584]:
```suggestion
        rustc_never_type_mode, Normal, template!(NameValueStr: "fallback_to_unit|fallback_to_niko|fallback_to_never|no_fallback"), ErrorFollowing,
```

**Comment by compiler-errors** [comment]
> There is a place where I'm not sure what to do:
> 
> https://github.com/rust-lang/rust/blob/c5b571310dc60c6a58c6505cce2fb7e2f3f9aa68/compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs#L396-L402
> 
> I have no context on what this does..

It's for when there's no opaque type inference. Consider something like:

```
#![feature(never_type_fallback)]

trait Uwu {}
impl Uwu for () {}

fn foo() -> impl Uwu { foo() }
//~^ ERROR the trait bound `!: Uwu` is not satisfied
```

**Review by compiler-errors — APPROVED** [review]
thanks waffle :heart: 

this is good enough for now though we may want to figure out wtf to do with that opaque type fallback :skull: 

also lol @ `fallback_to_niko`

**Pull request merged**